clang/lib/Sema/Metafunctions.cpp

```diff
@@ -1252,6 +1257,66 @@ bool map_decl_to_entity(APValue &Result, Sema &S, EvalFn Evaluator,
   llvm_unreachable("unknown reflection kind");
 }
 
+template <typename DT>
+bool getQualifiedName(APValue &Result, Sema &S, EvalFn Evaluator, DT& declType)
+{
+    SmallString<128> QualifiedNameBuffer;
+    llvm::raw_svector_ostream OS(QualifiedNameBuffer);
+    declType.printQualifiedName(OS);
+    StringRef Name = QualifiedNameBuffer.str();
+    if( makeCString(Result, Name, S.Context, Evaluator) )
+    {
+      llvm_unreachable("failed to create qualified name string");
+      return true;
+    }
+    return false;
+}
+
+bool qualified_name_of(APValue &Result, Sema &S, EvalFn Evaluator, QualType ResultTy,
+             SourceRange Range, ArrayRef<Expr *> Args) {
+  assert(Args[0]->getType()->isReflectionType());
+  assert(ResultTy ==
+         S.Context.getPointerType(S.Context.getConstType(S.Context.CharTy)));
+
+  APValue R;
+  if (!Evaluator(R, Args[0], true))
+    return true;
+
+  switch (R.getReflection().getKind()) {
+  case ReflectionValue::RK_type: { 
+    QualType QT = R.getReflectedType(); 
+    return SetAndSucceed(Result, getTypeName(S.Context, Evaluator, QT,
+                                             /*emptyIfUnnamed=*/true));
+  }
+  case ReflectionValue::RK_declaration: { 
+    const ValueDecl *VD = cast<ValueDecl>(R.getReflectedDecl());
+    return getQualifiedName(Result, S, Evaluator, *VD);
+  }
+  case ReflectionValue::RK_template: {  
+    const TemplateDecl *TD = cast<TemplateDecl>(R.getReflectedTemplate().getAsTemplateDecl());
+    return getQualifiedName(Result, S, Evaluator, *TD);
+  }
+  case ReflectionValue::RK_const_value: {
+    if (makeCString(Result, "", S.Context, Evaluator))
+      llvm_unreachable("failed to create empty string");
+    return false;
+  }
+  case ReflectionValue::RK_namespace: {
+    const NamespaceDecl *ND = cast<NamespaceDecl>(R.getReflectedNamespace());
+    return getQualifiedName(Result, S, Evaluator, *ND);
+  }
+  case ReflectionValue::RK_base_specifier: {
+    QualType QT = R.getReflectedBaseSpecifier()->getType();
+    const Decl* D = findTypeDecl(QT);
+    const TypeDecl *TD = cast<TypeDecl>(D);
+    return getQualifiedName(Result, S, Evaluator, *TD);
+  }
+  case ReflectionValue::RK_data_member_spec:
+    return true;
```

Comment:
Although this is what we do for `name_of`, it's not quite correct - P2996 specifies:
> For all other reflections, an empty string_view is produced.

So that should include this case as well. Let's just stack this case on top of `RK_const_value`.
```cpp
case ReflectionValue::RK_const_value:
case ReflectionValue::RK_data_member_spec: {
  if (makeCString(Result, "", S.Context, Evaluator))
    llvm_unreachable("failed to create empty string");
  return false;
}
```

(bonus points if you want to fix that in `name_of`, while you're in here 😄 )